### PR TITLE
[heft-jest-plugin] Upgrade Jest from `~27.4.2` to `~29.3.1`

### LINKED ISSUE #3440: [heft-jest-plugin] Upgrade to Jest 28.1.0+ to support additional ESM constructs
## Summary

Jest 27 has limited support for ES modules.  For example, it doesn't support subpath exports.  

From what I can see, Jest 28 now adds more support for ESM.  It'd be great if we can upgrade the jest libs here to use that.

### LINKED ISSUE #3665: [heft] upgrade or add jest 28/29 plugins
I usually don't question the pace of the rush repo but at the same time I am irked (might be irrational) by my heft-jest-plugin projects using jest 27.

Now that jest is on v29 I think we (I consider myself a rusher now (even if ya don't)) should aim to not be behind more than one major release. 

_PS: on a semi related note I should really find the time to PR in my tap and vitest heft plugins which (for me and my team) seem to work p well (just gotta clean them up (a bunch)). My team is often using a hybrid version of the heft jest plugin and my messy vitest plugin._

PPS/PSS: I really dislike jest but heft + jest is 🔥 so hats off to everyone involved.

### PR DESCRIPTION
**Pull request by tomi-bigpi**
## Summary
Upgrades heft-jest-plugin (and related rigs/packages) to use Jest `~29.3.1`.

Fixes #3665

Fixes #3440

## Details
* Jest packages were updated to use `~29.3.1`
* Required change to return object from transformers
* Update types to match current version, including `Context` to `TestContext`, `Config.Path` to `string`
* `snapshotFormat` configuration option was set to retain existing behavior to increase backwards compatibility
* `testURL` moved to new `testEnvironmentOptions.url` format

Note: There is a `@ts-expect-error` related to passing in reporters to `Config.argv` (`jestArgv`). The type on that is `string[]`, but works with the previous array of `JestReporterConfig[]` due to the direct call to the (non-public?) `runCLI` method in Jest.

## How it was tested
Ran various combinations of `rush purge`, `rush update`, `rush build`, `rush rebuild`, and `rush test`. Additionally, tested running tests in our internal repository with the changes. Was able to run Jest against ESM only code with `NODE_OPTIONS=--experimental-vm-modules` and  `"extensionsToTreatAsEsm": [".ts"]` config option per Jest's documentation.

Additionally tested with and without the `--debug-heft-reporter` option.

## Impacted documentation
No direct changes required, unless we want to highlight the need for adding an import to `@jest/globals` instead of directly referencing `jest` via the 3rd party global typings, or the optional modifications to `jest.config.json` files to support the new Jest `snapshotFormat` option, or add a mention to release notes for anyone that may not be using the `jest-shared.config.json` as the basis for their configuration.

For completeness, the following links have Jest's related upgrade documentation:
* https://jestjs.io/docs/28.x/upgrading-to-jest28
* https://jestjs.io/docs/upgrading-to-jest29

### CODE REVIEW

**Comment by tomi-bigpi** [comment]
@microsoft-github-policy-service agree [company="Big Pi Ventures, Inc."]

**Comment by tomi-bigpi** [comment]
@microsoft-github-policy-service agree company="Big Pi Ventures, Inc."

**Review by D4N14L — DISMISSED** [review]
Generally looks good, just a couple small changes.

File: build-tests/heft-jest-reporters-test/tsconfig.json [hunk 1]
```diff
@@ -17,7 +17,7 @@
 
     "module": "esnext",
     "moduleResolution": "node",
-    "target": "es5",
+    "target": "ES2015",
     "lib": ["es5"]
```
[the remark this answers is not in the document]
Comment (tomi-bigpi) [reply build-tests/heft-jest-reporters-test/tsconfig.json:21]: Sorry, not quite following. Did you mean to also bump `lib` to `ES2015` or something else?

File: rigs/heft-node-rig/package.json [hunk 2]
```diff
@@ -19,7 +19,8 @@
     "@microsoft/api-extractor": "workspace:*",
     "@rushstack/heft-jest-plugin": "workspace:*",
     "eslint": "~8.7.0",
-    "jest-environment-node": "~27.4.2",
+    "jest-environment-jsdom": "~29.3.1",
```
[the remark this answers is not in the document]
Comment (tomi-bigpi) [reply rigs/heft-node-rig/package.json]: Ah, I was following the Jest upgrade docs a little too closely. Will remove.

File: build-tests/heft-jest-reporters-test/tsconfig.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (D4N14L) [reply build-tests/heft-jest-reporters-test/tsconfig.json]: Yeah that's what I was saying. Having lib targeting an older version than the target is an odd situation. May as well keep them aligned.

**Comment by jessekrubin** [comment]
I have been waiting for this one!

**Comment by woss** [comment]
hi, there, this is a SUPER important feature! i was looking to see why is this still blocked. everything seems OK

**Review by iclanton — APPROVED** [review]

**Pull request merged**